File: clients/cpp/DebugInterface.cpp
#include "DebugInterface.hpp"
#include "codegame/ClientMessage.hpp"

DebugInterface::DebugInterface(TcpStream* stream): stream(stream) {}

void DebugInterface::addPlacedText(model::Vec2 position, std::string text, model::Vec2 alignment, double size, debugging::Color color)
{
    add(debugging::DebugData(debugging::PlacedText(position, text, alignment, size, color)));
}

void DebugInterface::addCircle(model::Vec2 position, double radius, debugging::Color color)
{
    add(debugging::DebugData(debugging::Circle(position, radius, color)));
}

void DebugInterface::addGradientCircle(model::Vec2 position, double radius, debugging::Color innerColor, debugging::Color outerColor)
{
    add(debugging::DebugData(debugging::GradientCircle(position, radius, innerColor, outerColor)));
}

void DebugInterface::addRing(model::Vec2 position, double radius, double width, debugging::Color color)
{
    add(debugging::DebugData(debugging::Ring(position, radius, width, color)));
}

void DebugInterface::addPie(model::Vec2 position, double radius, double startAngle, double endAngle, debugging::Color color)
{
    add(debugging::DebugData(debugging::Pie(position, radius, startAngle, endAngle, color)));
}

void DebugInterface::addArc(model::Vec2 position, double radius, double width, double startAngle, double endAngle, debugging::Color color)
{
    add(debugging::DebugData(debugging::Arc(position, radius, width, startAngle, endAngle, color)));
}

void DebugInterface::addRect(model::Vec2 bottomLeft, model::Vec2 size, debugging::Color color)
{
    add(debugging::DebugData(debugging::Rect(bottomLeft, size, color)));
}

void DebugInterface::addPolygon(std::vector<model::Vec2> vertices, debugging::Color color)
{
    add(debugging::DebugData(debugging::Polygon(vertices, color)));
}

void DebugInterface::addGradientPolygon(std::vector<debugging::ColoredVertex> vertices)
{
    add(debugging::DebugData(debugging::GradientPolygon(vertices)));
}

void DebugInterface::addSegment(model::Vec2 firstEnd, model::Vec2 secondEnd, double width, debugging::Color color)
{
    add(debugging::DebugData(debugging::Segment(firstEnd, secondEnd, width, color)));
}

void DebugInterface::addGradientSegment(model::Vec2 firstEnd, debugging::Color firstColor, model::Vec2 secondEnd, debugging::Color secondColor, double width)
{
    add(debugging::DebugData(debugging::GradientSegment(firstEnd, firstColor, secondEnd, secondColor, width)));
}

void DebugInterface::addPolyLine(std::vector<model::Vec2> vertices, double width, debugging::Color color)
{
    add(debugging::DebugData(debugging::PolyLine(vertices, width, color)));
}

void DebugInterface::addGradientPolyLine(std::vector<debugging::ColoredVertex> vertices, double width)
{
    add(debugging::DebugData(debugging::GradientPolyLine(vertices, width)));
}

void DebugInterface::add(debugging::DebugData debugData)
{
    send(debugging::DebugCommand(debugging::Add(debugData)));
}

void DebugInterface::clear()
{
    send(debugging::DebugCommand(debugging::Clear()));
}

void DebugInterface::setAutoFlush(bool enable)
{
    send(debugging::DebugCommand(debugging::SetAutoFlush(enable)));
}

void DebugInterface::flush()
{
    send(debugging::DebugCommand(debugging::Flush()));
}

void DebugInterface::send(debugging::DebugCommand command)
{
    codegame::ClientMessage message = codegame::DebugMessage(command);
    codegame::writeClientMessage(message, *stream);
    stream->flush();
}

debugging::DebugState DebugInterface::getState()
{
    codegame::ClientMessage message = codegame::RequestDebugState();
    codegame::writeClientMessage(message, *stream);
    stream->flush();
    return debugging::DebugState::readFrom(*stream);
}

File: clients/php/DebugInterface.php
<?php

require_once 'Codegame/ClientMessage.php';
require_once 'Debugging/DebugCommand.php';
require_once 'Debugging/DebugState.php';

class DebugInterface
{
    private TcpStream $tcpStream;

    function __construct(TcpStream $tcpStream)
    {
        $this->tcpStream = $tcpStream;
    }

    public function send(\Debugging\DebugCommand $command): void
    {
        (new \Codegame\ClientMessage\DebugMessage($command))->writeTo($this->tcpStream->outputStream);
        $this->tcpStream->outputStream->flush();
    }

    public function getState(): \Debugging\DebugState
    {
        (new \Codegame\ClientMessage\RequestDebugState())->writeTo($this->tcpStream->outputStream);
        $this->tcpStream->outputStream->flush();
        return \Debugging\DebugState\readFrom($this->tcpStream->inputStream);
    }

    function addPlacedText(\Model\Vec2 $position, string $text, \Model\Vec2 $alignment, float $size, \Debugging\Color $color): void
    {
        $this->add(new \Debugging\DebugData\PlacedText($position, $text, $alignment, $size, $color));
    }
    function addCircle(\Model\Vec2 $position, float $radius, \Debugging\Color $color): void
    {
        $this->add(new \Debugging\DebugData\Circle($position, $radius, $color));
    }
    function addGradientCircle(\Model\Vec2 $position, float $radius, \Debugging\Color $innerColor, \Debugging\Color $outerColor): void
    {
        $this->add(new \Debugging\DebugData\GradientCircle($position, $radius, $innerColor, $outerColor));
    }
    function addRing(\Model\Vec2 $position, float $radius, float $width, \Debugging\Color $color): void
    {
        $this->add(new \Debugging\DebugData\Ring($position, $radius, $width, $color));
    }
    function addPie(\Model\Vec2 $position, float $radius, float $startAngle, float $endAngle, \Debugging\Color $color): void
    {
        $this->add(new \Debugging\DebugData\Pie($position, $radius, $startAngle, $endAngle, $color));
    }
    function addArc(\Model\Vec2 $position, float $radius, float $width, float $startAngle, float $endAngle, \Debugging\Color $color): void
    {
        $this->add(new \Debugging\DebugData\Arc($position, $radius, $width, $startAngle, $endAngle, $color));
    }
    function addRect(\Model\Vec2 $bottomLeft, \Model\Vec2 $size, \Debugging\Color $color): void
    {
        $this->add(new \Debugging\DebugData\Rect($bottomLeft, $size, $color));
    }
    function addPolygon(array $vertices, \Debugging\Color $color): void
    {
        $this->add(new \Debugging\DebugData\Polygon($vertices, $color));
    }
    function addGradientPolygon(array $vertices): void
    {
        $this->add(new \Debugging\DebugData\GradientPolygon($vertices));
    }
    function addSegment(\Model\Vec2 $firstEnd, \Model\Vec2 $secondEnd, float $width, \Debugging\Color $color): void
    {
        $this->add(new \Debugging\DebugData\Segment($firstEnd, $secondEnd, $width, $color));
    }
    function addGradientSegment(\Model\Vec2 $firstEnd, \Debugging\Color $firstColor, \Model\Vec2 $secondEnd, \Debugging\Color $secondColor, float $width): void
    {
        $this->add(new \Debugging\DebugData\GradientSegment($firstEnd, $firstColor, $secondEnd, $secondColor, $width));
    }
    function addPolyLine(array $vertices, float $width, \Debugging\Color $color): void
    {
        $this->add(new \Debugging\DebugData\PolyLine($vertices, $width, $color));
    }
    function addGradientPolyLine(array $vertices, float $width): void
    {
        $this->add(new \Debugging\DebugData\GradientPolyLine($vertices, $width));
    }

    function add(\Debugging\DebugData $debugData): void
    {
        $this->send(new \Debugging\DebugCommand\Add($debugData));
    }
    function clear(): void
    {
        $this->send(new \Debugging\DebugCommand\Clear());
    }
    function setAutoFlush(bool $enable): void
    {
        $this->send(new \Debugging\DebugCommand\SetAutoFlush($enable));
    }
    function flush(): void
    {
        $this->send(new \Debugging\DebugCommand\Flush());
    }
}